api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -130,22 +155,32 @@ public Workspace get(String ns, String firecloudName) {
         .collect(Collectors.toList());
   }
 
+  @Transactional
   @Override
-  public List<WorkspaceResponse> getWorkspaces() {
-    return getWorkspacesAndPublicWorkspaces().stream()
-        .filter(
-            workspaceResponse ->
-                !(workspaceResponse.getAccessLevel() != WorkspaceAccessLevel.OWNER
-                    && workspaceResponse.getAccessLevel() != WorkspaceAccessLevel.WRITER
-                    && workspaceResponse.getWorkspace().getPublished()))
-        .collect(Collectors.toList());
-  }
+  public WorkspaceResponse getWorkspace(String workspaceNamespace, String workspaceId) {
+    Workspace dbWorkspace = getRequired(workspaceNamespace, workspaceId);
 
-  @Override
-  public List<WorkspaceResponse> getPublishedWorkspaces() {
-    return getWorkspacesAndPublicWorkspaces().stream()
-        .filter(workspaceResponse -> workspaceResponse.getWorkspace().getPublished())
-        .collect(Collectors.toList());
+    org.pmiops.workbench.firecloud.model.WorkspaceResponse fcResponse;
+    org.pmiops.workbench.firecloud.model.Workspace fcWorkspace;
+
+    WorkspaceResponse response = new WorkspaceResponse();
+
+    // This enforces access controls.
+    fcResponse = fireCloudService.getWorkspace(workspaceNamespace, workspaceId);
+    fcWorkspace = fcResponse.getWorkspace();
+
+    if (fcResponse.getAccessLevel().equals(WorkspaceService.PROJECT_OWNER_ACCESS_LEVEL)) {
+      // We don't expose PROJECT_OWNER in our API; just use OWNER.
+      response.setAccessLevel(WorkspaceAccessLevel.OWNER);
+    } else {
+      response.setAccessLevel(WorkspaceAccessLevel.fromValue(fcResponse.getAccessLevel()));
+      if (response.getAccessLevel() == null) {
+        throw new ServerErrorException("Unsupported access level: " + fcResponse.getAccessLevel());
```

Comment:
Do you really want to throw here, or just put the error type in `WorkspaceResponse` and return it?

Reply:
I just moved this over from `WorkspaceController` so I'm a little hesitant to make functional changes. It could be that this condition is enforcing some access control.

Reply:
Might be a conversation for a later date. One way to look at it is that a 500 should really get someone's attention if they're monitoring things, so if this situation can happen in normal UI or API operations, then a 403 exception would be better I would think.

Reply:
This line corresponds to a programming error or unexpected response from Firecloud, so 500 is probably appropriate. Any expected 403s would have propagated from line L162

Reply:
Thanks CH. So is Unsupported Access level still what we want?